README.rst
.. image:: https://readthedocs.org/projects/pymysql/badge/?version=latest
    :target: https://pymysql.readthedocs.io/
    :alt: Documentation Status

.. image:: https://badge.fury.io/py/PyMySQL.svg
    :target: https://badge.fury.io/py/PyMySQL

.. image:: https://travis-ci.org/PyMySQL/PyMySQL.svg?branch=master
    :target: https://travis-ci.org/PyMySQL/PyMySQL

.. image:: https://coveralls.io/repos/PyMySQL/PyMySQL/badge.svg?branch=master&service=github
    :target: https://coveralls.io/github/PyMySQL/PyMySQL?branch=master

.. image:: https://img.shields.io/badge/license-MIT-blue.svg
    :target: https://github.com/PyMySQL/PyMySQL/blob/master/LICENSE


PyMySQL
=======

.. contents:: Table of Contents
   :local:

This package contains a pure-Python MySQL client library, based on `PEP 249`_.

Most public APIs are compatible with mysqlclient and MySQLdb.

NOTE: PyMySQL doesn't support low level APIs `_mysql` provides like `data_seek`,
`store_result`, and `use_result`. You should use high level APIs defined in `PEP 249`_.
But some APIs like `autocommit` and `ping` are supported because `PEP 249`_ doesn't cover
their usecase.

.. _`PEP 249`: https://www.python.org/dev/peps/pep-0249/


Requirements
-------------

* Python -- one of the following:

  - CPython_ : 2.7 and >= 3.5
  - PyPy_ : Latest version

* MySQL Server -- one of the following:

  - MySQL_ >= 5.5
  - MariaDB_ >= 5.5

.. _CPython: https://www.python.org/
.. _PyPy: https://pypy.org/
.. _MySQL: https://www.mysql.com/
.. _MariaDB: https://mariadb.org/


Installation
------------

Package is uploaded on `PyPI <https://pypi.org/project/PyMySQL>`_.

You can install it with pip::

    $ python3 -m pip install PyMySQL

To use "sha256_password" or "caching_sha2_password" for authenticate,
you need to install additional dependency::

   $ python3 -m pip install PyMySQL[rsa]


Documentation
-------------

Documentation is available online: https://pymysql.readthedocs.io/

For support, please refer to the `StackOverflow
<https://stackoverflow.com/questions/tagged/pymysql>`_.

Example
-------

The following examples make use of a simple table

.. code:: sql

   CREATE TABLE `users` (
       `id` int(11) NOT NULL AUTO_INCREMENT,
       `email` varchar(255) COLLATE utf8_bin NOT NULL,
       `password` varchar(255) COLLATE utf8_bin NOT NULL,
       PRIMARY KEY (`id`)
   ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
   AUTO_INCREMENT=1 ;


.. code:: python

    import pymysql.cursors

    # Connect to the database
    connection = pymysql.connect(host='localhost',
                                 user='user',
                                 password='passwd',
                                 db='db',
                                 charset='utf8mb4',
                                 cursorclass=pymysql.cursors.DictCursor)

    try:
        with connection.cursor() as cursor:
            # Create a new record
            sql = "INSERT INTO `users` (`email`, `password`) VALUES (%s, %s)"
            cursor.execute(sql, ('webmaster@python.org', 'very-secret'))

        # connection is not autocommit by default. So you must commit to save
        # your changes.
        connection.commit()

        with connection.cursor() as cursor:
            # Read a single record
            sql = "SELECT `id`, `password` FROM `users` WHERE `email`=%s"
            cursor.execute(sql, ('webmaster@python.org',))
            result = cursor.fetchone()
            print(result)
    finally:
        connection.close()

This example will print:

.. code:: python

    {'password': 'very-secret', 'id': 1}


Resources
---------

* DB-API 2.0: https://www.python.org/dev/peps/pep-0249/

* MySQL Reference Manuals: https://dev.mysql.com/doc/

* MySQL client/server protocol:
  https://dev.mysql.com/doc/internals/en/client-server-protocol.html

* "Connector" channel in MySQL Community Slack:
  https://lefred.be/mysql-community-on-slack/

* PyMySQL mailing list: https://groups.google.com/forum/#!forum/pymysql-users

License
-------

PyMySQL is released under the MIT License. See LICENSE for more information.


CHANGELOG
# Changes

## 0.9.3

Release date: 2018-12-18

* cryptography dependency is optional now.
* Fix old_password (used before MySQL 4.1) support.
* Deprecate old_password.
* Stop sending ``sys.argv[0]`` for connection attribute "program_name".
* Close connection when unknown error is happened.
* Deprecate context manager API of Connection object.

## 0.9.2

Release date: 2018-07-04

* Disabled unintentinally enabled debug log
* Removed unintentionally installed tests


## 0.9.1

Release date: 2018-07-03

* Fixed caching_sha2_password and sha256_password raise TypeError on PY2
  (#700, #702)


## 0.9.0

Release date: 2018-06-27

* Change default charset from latin1 to utf8mb4.  (because MySQL 8 changed) (#692)
* Support sha256_password and caching_sha2_password auth method (#682)
* Add cryptography dependency, because it's needed for new auth methods.
* Remove deprecated `no_delay` option (#694)
* Support connection attributes (#679)
* Map LOCK_DEADLOCK to OperationalError (#693)

## 0.8.1

Release date: 2018-05-07

* Reduce `cursor.callproc()` roundtrip time. (#636)

* Fixed `cursor.query()` is hunged after multi statement failed. (#647)

* WRONG_DB_NAME and WRONG_COLUMN_NAME is ProgrammingError for now. (#629)

* Many test suite improvements, especially adding MySQL 8.0 and using Docker.
  Thanks to Daniel Black.

* Dropped support for old Python and MySQL which is not tested long time.


## 0.8

Release date: 2017-12-20

* **BACKWARD INCOMPATIBLE** ``binary_prefix`` option is added and off
  by default because of compatibility with mysqlclient.
  When you need PyMySQL 0.7 behavior, you have to pass ``binary_prefix=True``.
  (#549)

* **BACKWARD INCOMPATIBLE** ``MULTI_STATEMENTS`` client flag is no longer
  set by default, while it was on PyMySQL 0.7.  You need to pass
  ``client_flag=CLIENT.MULTI_STATEMENTS`` when you connect to explicitly
  enable multi-statement mode. (#590)

* Fixed AuthSwitch packet handling.

* Raise OperationalError for MariaDB's constraint error. (#607)

* executemany() accepts query without space between ``VALUES`` and ``(``.  (#597)

* Support config file containing option without value. (#588)

* Fixed Connection.ping() returned unintended value.


## 0.7.11

Release date: 2017-04-06

* Fixed Connection.close() failed when failed to send COM_CLOSE packet.
* Cursor.executemany() accepts query ends with semicolon.
* ssl parameters can be read from my.cnf.


## 0.7.10

Release date: 2017-02-14

* **SECURITY FIX**: Raise RuntimeError when received LOAD_LOCAL packet while
  ``loacal_infile=False``.  (Thanks to Bryan Helmig)

* Raise SERVER_LOST error for MariaDB's shutdown packet (#540)

* Change default connect_timeout to 10.

* Add bind_address option (#529)


## 0.7.9

Release date: 2016-09-03

* Fix PyMySQL stop reading rows when first column is empty string (#513)
  Reverts DEPRECATE_EOF introduced in 0.7.7.

## 0.7.8

Release date: 2016-09-01

* Revert error message change in 0.7.7.
  (SQLAlchemy parses error message, #507)

## 0.7.7

Release date: 2016-08-30

* Add new unicode collation (#498)
* Fix conv option is not used for encoding objects.
* Experimental support for DEPRECATE_EOF protocol.

## 0.7.6

Release date: 2016-07-29

* Fix SELECT JSON type cause UnicodeError
* Avoid float convertion while parsing microseconds
* Warning has number
* SSCursor supports warnings

## 0.7.5

Release date: 2016-06-28

* Fix exception raised while importing when getpwuid() fails (#472)
* SSCursor supports LOAD DATA LOCAL INFILE (#473)
* Fix encoding error happen for JSON type (#477)
* Fix test fail on Python 2.7 and MySQL 5.7 (#478)

## 0.7.4

Release date: 2016-05-26

* Fix AttributeError may happen while Connection.__del__ (#463)
* Fix SyntaxError in test_cursor. (#464)
* frozenset support for query value. (#461)
* Start using readthedocs.io

## 0.7.3

Release date: 2016-05-19

* Add read_timeout and write_timeout option.
* Support serialization customization by `conv` option.
* Unknown type is converted by `str()`, for MySQLdb compatibility.
* Support '%%' in `Cursor.executemany()`
* Support REPLACE statement in `Cursor.executemany()`
* Fix handling incomplete row caused by 'SHOW SLAVE HOSTS'.
* Fix decode error when use_unicode=False on PY3
* Fix port option in my.cnf file is ignored.


## 0.7.2

Release date: 2016-02-24

* Fix misuse of `max_allowed_packet` parameter. (#426, #407 and #397)
* Add %(name)s plceholder support to `Cursor.executemany()`. (#427, thanks to
  @WorldException)

## 0.7.1

Release date: 2016-01-14

* Fix auth fail with MySQL 5.1
* Fix escaping unicode fails on Python 2

## 0.7

Release date: 2016-01-10

* Faster binary escaping
* Add `"_binary" prefix` to string literal for binary types.
  binary types are: `bytearray` on Python 2, `bytes` and `bytearray` on Python 3.
  This is because recent MySQL show warnings when string literal is invalid for
  connection encoding.
* `pymysql.Binary()` returns `bytearray` on Python 2.  This is required to distinguish
  binary and string.
* Auth plugin support.
* no_delay option is ignored.  It will be removed in PyMySQL 0.8.


## 0.6.7

Release date: 2015-09-30

* Allow self signed certificate
* Add max_allowed_packet option
* Fix error when bytes in executemany
* Support geometry type
* Add coveralls badge to README
* Fix some bugs relating to warnings
* Add Cursor.mogrify() method
* no_delay option is deprecated and True by default
* Fix options from my.cnf overrides options from arguments
* Allow socket like object.  (It's not feature for end users)
* Strip quotes while reading options from my.cnf file
* Fix encoding issue in executemany()

## 0.6.6

* Add context manager to cursor
* Fix can't encode blob that is not utf-8 on PY3. (regression of 0.6.4,
  Thanks to @wiggzz)

## 0.6.5

draft2.py
# # This file contains functions to combine data from different sources together.

# import pandas as pd
# import QUANTAXIS as qa
# import pymongo
# import json
# import datetime


# def readdata(code, start, end):
#     '''This function loads data from local mongodb using qafetch function from quantaxis.
#         Args:
#             code: stock code, string type.
#             start: start date.
#             end: end date.
#         Returns:
#             A data frame indexed with standard datetime index.
#     '''
#     data = qa.QA_fetch_stock_day_adv(code=code, start=start, end=end)
#     result = data.data
#     result = result.sort_index(ascending=False)
#     result = result.reset_index(level=1)
#     result = result.drop(columns='code')

#     return result


# def get_dtindex(start, end):
#     dtindex = pd.date_range(start=start,
#                             end=end,
#                             freq='D')
#     return dtindex


# def get_parameters(code, start, end):
#     outcome = readdata(code=code, start=start, end=end)
#     dtindex = get_dtindex(start=start, end=end)
#     return outcome, dtindex


# def upload_data(code, start, end):
#     outcome, dtindex = get_parameters(code=code, start=start, end=end)
#     # connect to mongodb
#     myclient = pymongo.MongoClient("mongodb://localhost:27017/")
#     database = myclient['mydatabase']
#     datacol = database[code + str(datetime.date.today())]
#     dtindexcol = database[code + str(datetime.date.today()) + 'index']
#     outcome['date'] = outcome.index

#     # datacol.insert(json.loads(outcome.T.to_json()).values())
#     # dtindexcol.insert(json.loads(outcome.T.to_json()).values())
#     return outcome


# def QA_util_to_json_from_pandas(data):
#     """需要对于datetime 和date 进行转换, 以免直接被变成了时间戳"""
#     # if 'datetime' in data.columns:
#     #     data.datetime = data.datetime.apply(str)
#     if 'date' in data.columns:
#         data.date = data.date.apply(str)
#     return json.loads(data.to_json(orient='records'))
#     # return data


# code = '000002'
# start = '2009-07-13'
# end = '2019-07-12'

# myclient = pymongo.MongoClient("mongodb://localhost:27017/")
# database = myclient['mydatabase']
# datacol = database[code + str(datetime.date.today())]

# dtindex = pd.date_range(start=start, end=end, freq='D')
# outcome = upload_data(code=code, start=start, end=end)
# outcome = QA_util_to_json_from_pandas(outcome)
# # print(outcome)
# datacol.insert_many(outcome)
# # result = json.loads(outcome.to_json(orient='records'))
# # result = result.decode(encoding='utf-8')
# # print(result)
# # result2 = json.loads(result)
# # print(result2)


# from QUANTAXIS.TSBoosting.TSBoosting import TS_Boosting_predict
# start = '2007-10-18'
# end = '2019-08-18'
# by = 'D'
# TS_Boosting_predict(start=start, end=end, by=by, databaseid='mydatabase', collectionid='rawdatatest')


# import os
# import pymongo
# from bson.json_util import dumps

# client = pymongo.MongoClient("mongodb://10.0.75.1:27017/")
# change_stream = client.mydatabase.rawdatatest.watch()
# for change in change_stream:
# import time
# import subprocess


# def run_sleep(period):
#     proc = subprocess.Popen(['timeout', '/t', str(period)])
#     return proc


# start = time.time()
# procs = []
# for _ in range(2):
#     print('start')
#     proc = run_sleep(1)
#     print('end')
#     procs.append(proc)
#     print(proc)
# for proc in procs:
#     proc.communicate()
# end = time.time()
# print(end - start)


from datetime import datetime
import pandas as pd
# timestamp = 42000
# dt_object = datetime.fromtimestamp(timestamp)
# print(dt_object)
data = pd.read_csv('C:/Users/shixu/Downloads/trace_201708/container_usage.csv',
                   index_col=0, header=None)
print(data.head())
outcome1 = data.iloc[:2]
outcome2 = data.iloc[:3]
outcome1.index.names = ['datetime']
outcome1.rename(columns={1: 'y'})
outcome2.index.names = ['datetime']
outcome2.rename(columns={1: 'y'})
# outcome1['datetime'] = pd.to_datetime(outcome1['datetime'])
# outcome2['datetime'] = pd.to_datetime(outcome2['datetime'])

print(outcome1)
print(outcome2)
